packages/text-field/Input.js

```diff
@@ -0,0 +1,220 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+// whitelist based off of https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5/Constraint_validation
+// under section: `Validation-related attributes`
+const VALIDATION_ATTR_WHITELIST = [
+  'pattern', 'min', 'max', 'required', 'step', 'minlength', 'maxlength',
+];
+
+export default class Input extends React.Component {
+  constructor(props) {
+    super(props);
+    this.inputElement = React.createRef();
+    props.setBadInputHandler(this.badInputHandler);
+    props.setIsValidHandler(this.isValidHandler);
+    this.state = {
+      value: props.value,
+    };
+  }
+
+  badInputHandler = () => this.inputElement.current.validity.badInput;
+  isValidHandler = () => this.inputElement.current.validity.valid;
+
+  get classes() {
+    return classnames('mdc-text-field__input', this.props.className);
+  }
+
+  componentDidMount() {
+    this.props.handleValueChange(this.props.value);
+    if (this.props.id) {
+      this.props.setInputId(this.props.id);
+    }
+    if (this.props.disabled) {
+      this.props.setDisabled(true);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const {foundation} = nextProps;
+    this.handleValidationAttributeUpdate(nextProps);
+
+    if (this.props.disabled !== nextProps.disabled) {
+      nextProps.setDisabled(nextProps.disabled);
+      foundation.setDisabled(nextProps.disabled);
+    }
+
+    if (this.props.id !== nextProps.id) {
+      nextProps.setInputId(nextProps.id);
+    }
+
+    this.updateValue(nextProps);
+  }
+
+  handleValidationAttributeUpdate = (nextProps) => {
+    const {foundation} = nextProps;
+    VALIDATION_ATTR_WHITELIST.some((attr) => {
```

Comment:
This will only update the first prop that changed and ignore the others. Is that what we want?

Reply:
Yes that's right. Its actually similar to what we do in MDC-Web. https://github.com/material-components/material-components-web/blob/master/packages/mdc-textfield/foundation.js#L187

Reply:
Awesome, thanks for the explanation!